Chartbeat Utilities

12 opened the pull request:
Resolves #1986 

**Overall change:** Add helper functions to aid Chartbeat implementation.

**Code changes:**

- Add new file `chartbeat.js` containing helpers to gather required info for Chartbeat.
- Add corresponding tests

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval

12 committed:
Build basic chartbeat func file + tests

12 committed:
Tidy up chartbeat utils

12 committed:
Merge branch 'latest' of github.com:bbc/simorgh into 1986_chartbeat

12 committed:
Refactor domain constructor function, begin sections constructor

12 committed:
Tests for Chartbeat domain

12 committed:
Merge branch 'latest' of github.com:bbc/simorgh into 1986_chartbeat

12 committed:
Test pageType and basic tests for sections

12 committed:
Update sections function to generate sections property

sareh commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,72 @@
+
+  describe('Chartbeat Domains', () => {
+    const services = [
+      ['news', 'bbc.co.uk'],
```
To make the test below more readable, I'd suggest changing this to an array of objects. It is more verbose, but clearer what the list contains at a glance. I'd suggest doing the same for the other tests in this file too. 
```
const domains = [
  { service: 'news', expectedDomain: 'bbc.co.uk' }, 
  { service: 'persian', expectedDomain: 'persian.bbc.co.uk' }, 
....
];
```
So then you can have the expectation to be clearer. 
`expect(domain(service)).toBe(expectedDomain);`

sareh commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,39 @@
+
+export const useCanonical = true;
+
+export const cknsSylphid = () => {};
```
Explaining what this const name is would add clarity.

sareh commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,39 @@
+};
+
+export const type = pageType =>
+  pageType === 'article' ? 'News Article' : 'Index';
```
```suggestion
  pageType === 'article' ? 'New Article' : 'Index';
```
To match the requirements in the issue.

sareh commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,39 @@
+
+const buildServiceType = ({ service }, type) => [`${service} - ${type}`];
+
+export const sections = x => {
```
I'd suggest passing in pageType, service, producer and chapter at the top level, rather than nesting them in an object x. 

```suggestion
export const sections = (pageType, service, chapter = '', producer = '', ) => {
```

sareh reviewed (commented):
Great work on getting the functionality that is required. I've added some comments to improve the readability and hopefully simplify the code.

12 committed:
Ease readability for helpers and their corresponding tests

12 committed:
Remove redundant spread

12 committed:
Refactor sections constructor + make tests more readable

12 committed:
Add helper func to get sylphid cookie + corresponding tests

12 committed:
Merge branch 'latest' into 1986_chartbeat

dr3 commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,45 @@
+    return null;
+  }
+
+  return Cookie.get(ID_COOKIE) ? Cookie.get(ID_COOKIE) : null;
```
```suggestion
  return Cookie.get(ID_COOKIE) || null;
```

pjlee11 replied:
```suggestion
  return ? Cookie.get(ID_COOKIE) || null;
```

Does this work?

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,45 @@
+  return Cookie.get(ID_COOKIE) ? Cookie.get(ID_COOKIE) : null;
+};
+
+export const domain = service => {
```
```suggestion
export const getDomain = service => {
```

And change the references to this function

12 committed:
Use or return, and rename domain function to getDomain

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,45 @@
+  return parts.join(', ');
+};
+
+export const type = pageType =>
```
```suggestion
export const getType = pageType =>
```

Rename and just highlighting that this doesn't scale to more pageTypes but we can handle that in the future.

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,45 @@
+
+const buildServiceType = (service, type) => `${service} - ${type}`;
+
+export const sections = (service, type, producer, chapter) => {
```
```suggestion
export const buildSections = (service, type, producer, chapter) => {
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,135 @@
+  getSylphidCookie,
+  getDomain,
+  sections,
+  type,
```
```suggestion
  getType,
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,135 @@
+  useCanonical,
+  getSylphidCookie,
+  getDomain,
+  sections,
```
```suggestion
  buildSections,
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,135 @@
+
+    types.forEach(({ type: rawType, expectedType }) => {
+      it(`Type ${rawType} should return ${expectedType}`, () => {
+        expect(type(rawType)).toBe(expectedType);
```
```suggestion
        expect(getType(rawType)).toBe(expectedType);
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,135 @@
+    sectionFixtures.forEach(
+      ([{ service, producer, chapter, pageType, expected }]) => {
+        it(`sections should return "${expected}"`, () => {
+          expect(sections(service, pageType, producer, chapter)).toBe(expected);
```
```suggestion
          expect(buildSections(service, pageType, producer, chapter)).toBe(expected);
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,135 @@
+
+    sectionFixtures.forEach(
+      ([{ service, producer, chapter, pageType, expected }]) => {
+        it(`sections should return "${expected}"`, () => {
```
```suggestion
        it(`buildSections should return "${expected}"`, () => {
```

12 committed:
Phil being picky

Co-Authored-By: Phil Lee <phil.lee.ext@bbc.co.uk>

12 committed:
Update src/app/lib/analyticsUtils/chartbeat.test.js

Co-Authored-By: Phil Lee <phil.lee.ext@bbc.co.uk>

12 committed:
Merge branch 'latest' into 1986_chartbeat

12 committed:
Fix lint

12 committed:
Merge branch 'latest' into 1986_chartbeat

12 committed:
Merge branch 'latest' into 1986_chartbeat

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,137 @@
+    });
+
+    it('should return the contents of the ID cookie when a value is present', () => {
+      const expectedCookie = 'foobar';
```
```suggestion
      const expectedCookieValue = 'foobar';
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,137 @@
+
+      Object.defineProperty(window.document, 'cookie', {
+        writable: true,
+        value: `ckns_sylphid=${expectedCookie}`,
```
```suggestion
        value: `ckns_sylphid=${expectedCookieValue}`,
```

pjlee11 commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,137 @@
+        value: `ckns_sylphid=${expectedCookie}`,
+      });
+
+      expect(getSylphidCookie()).toBe(expectedCookie);
```
```suggestion
      expect(getSylphidCookie()).toBe(expectedCookieValue);
```

12 committed:
expectedCookie => expectedCookieValue

Co-Authored-By: Phil Lee <phil.lee.ext@bbc.co.uk>

12 committed:
Merge branch 'latest' into 1986_chartbeat

pjlee11 reviewed (approved):
LGTM 👍 

I've asked @jimjohnsonrollings to check the `expected` values in the tests from a requirements POV

sareh commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,137 @@
+        },
+      ],
+      [
+        {
```
Can you please add two more test cases:
- having a producer but not a chapter
- having a chapter but not a producer

Looking at the issue, there are these cases to be captured (note that `News` is camel-cased)

- If service is News and the ATI Producer is News: "News"
- If service is News and the ATI Producer is not News: "News - {Producer name in lowercase}" e.g. "News - wales"
- If the service is News and there is an ATI Chapter, then "News - {Final part of chapter in lower case}" - e.g. "News - africa" Then repeat the same entries again appended with " - ART" 
- Homepages - second value should be the name of the service appended with " - IDX"

sareh commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,137 @@
+
+    sectionFixtures.forEach(
+      ([{ service, producer, chapter, pageType, expected }]) => {
+        it(`buildSections should return "${expected}"`, () => {
```
It'd be good to update the description to explain the output. e.g.
```suggestion
        it(description, () => {
```
with in the `sectionFixtures` having `description` fields explaining what's happening:

e.g. `description: 'should add chapter if passed in',`

sareh reviewed (commented):
Looks great. Just some additional tests required & ensuring the case of the section text matches that in the Issue description.

12 committed:
Update sections constructor to cover all cases described

Update tests to reflect and test these additional cases
Add unique descriptions for each sections test for additional context

12 committed:
Merge branch 'latest' of github.com:bbc/simorgh into 1986_chartbeat

12 committed:
Merge branch 'latest' into 1986_chartbeat

jamesdonoh commented:
I know the issue says _Integrating with Chartbeat will be in another ticket._ but in order to clarify the purpose of this work, please could you you provide a rough example of how the functions will be used in practice? Just trying to gauge whether what we're exporting is at the right level of abstraction.

jamesdonoh commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,158 @@
+        value: `ckns_sylphid=${expectedCookieValue}`,
+      });
+
+      expect(getSylphidCookie()).toBe(expectedCookieValue);
```
We're testing `js-cookie` behaviour here rather than our own logic, no?

12 replied:
This series of tests test the following:

- `onClient` is `false` (the first branch of the function under test)
- `onClient` is `true`, and `ckns_sylphid` returns a truthy value (the second branch of the function under test)
- `onClient` is `true`, but `ckns_sylphid` returns a falsy value (the third branch of the function under test)

It does this by mocking the response of `document.cookie`, so yes, it's testing both the `getSylphidCookie` method, but also `js-cookie`, which probably is unnecessary looking back on it, but shouldn't cause us any issues?

Happy to change this to mock the response of `Cookie.get` if you'd prefer to ensure we're just testing this single method🙂

jamesdonoh replied:
Thanks for the summary. Yeah, I guess I would just have mocked `Cookie.get` but it works and can't see it causing any issues, happy to leave as-is.

12 replied:
Yeah so would have I looking back on it... not sure why I didn't do it that way to begin with. I've changed the test to reflect this.

jamesdonoh commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,48 @@
+    return null;
+  }
+
+  return Cookie.get(ID_COOKIE) || null;
```
According to docs `Cookie.get` can return undefined, is there a reason we want to convert anything falsy to `null`?

12 replied:
Good spot. This tidies the function up nicely

jamesdonoh commented on src/app/lib/analyticsUtils/chartbeat.test.js:
```diff
@@ -0,0 +1,158 @@
+      {
+        type: 'index',
+        expectedType: 'Index',
+      },
```
I'd like another case to confirm intended behaviour when `type` is neither.

12 replied:
Have implemented

jamesdonoh commented on src/app/lib/analyticsUtils/chartbeat.js:
```diff
@@ -0,0 +1,48 @@
+
+const buildServiceType = (service, type) => `${service} - ${type}`;
+
+const capitalize = s => s.charAt(0).toUpperCase() + s.slice(1);
```
Wondered if we should start a new library for generic text utils to encourage reuse and reduce the chance of someone duplicating this elsewhere later? There's nothing Chartbeat-specific about it.

12 replied:
I agree, something for this issue or a future one?

jamesdonoh replied:
Later is fine.

12 committed:
Tidy sylphid cookie method

jamesdonoh reviewed (commented):
Few belated thoughts, looks good in general though.

12 commented:
> I know the issue says _Integrating with Chartbeat will be in another ticket._ but in order to clarify the purpose of this work, please could you you provide a rough example of how the functions will be used in practice? Just trying to gauge whether what we're exporting is at the right level of abstraction.

@jamesdonoh, We have a config object that we need to set, that will be populated using the methods created in this PR. 

_all of the code below is untested and theoretical_

```js
import { ServiceContext } from '../../contexts/ServiceContext';
import { RequestContext } from '../../contexts/RequestContext';
import { getPromoHeadline } from '../../lib/analyticsUtils/article';
import * as cb from './chartbeat';

const { service } = useContext(ServiceContext);
const { pageType }  = useContext(RequestContext);
```

The above will set us up with the parameters that are required. 
Note that `producers` and `chapters` have not been fully configured and implemented yet, and so we should not send anything for these properties, following suit of how ATi currently works, as per:
```
Jim J-R [10:44]
We should take the same approach then for Chartbeat
```

Chartbeat then requires an object to be configured using the methods that this PR creates:
```js
var _sf_async_config = window._sf_async_config = (window._sf_async_config || {});

_sf_async_config.uid = cb.chartbeatUID; 
_sf_async_config.domain = cb.getDomain(service); 
_sf_async_config.useCanonical = cb.useCanonical;
_sf_async_config.sections = cb.buildSections(service, pageType, null, null); 
_sf_async_config.idSync.bbc_hid = cb.getSylphidCookie();
_sf_async_config.title = getPromoHeadline();
_sf_async_config.type = cb.getType();
```

The producer/chapter functionality I've added so that it should be instantly available when we can support those two properties.

jamesdonoh commented:
OK that helps, thanks. I was wondering if all these helpers actually needed to be exported separately but makes sense.

12 committed:
Refactor getType to allow return of shorthand types and account for front pages

- Update section tests to give appropriate result when type is null
- Update buildSections to return appropriate section when type is null
- Change cookie test to stub js-cookie, and not document.cookie (as the previous method was also testing js-cookie)

12 committed:
Merge branch 'latest' of github.com:bbc/simorgh into 1986_chartbeat

12 committed:
Merge branch 'latest' into 1986_chartbeat

12 committed:
Improve testing coverage on chartbeat

jamesdonoh reviewed (approved):
It's a 👍 from me.

jamesbrumpton commented:
LGTM
